### PR TITLE
Hivelord ability : Create Necromorph

### PR DESCRIPTION
**Pull request by Ketrai**
## About The Pull Request
Adds the ability for hivelords to sting dead and unrevivable human corpses to infect them with a necromorph virus. Necromorphs are squishier than normal zombies. Cannot regenerate, instead get gibbed upon death. Heal just as slowly. Are faster than normal zombies. Do not infect on hit unlike normal zombies. Don't take as long to transform corpses into zombies. Necromorphs inherit most of the zombie code, so they can be possessed like zombies when they spawn, and should give the player turned dibs if they're still in their body.

Uses the same snowflake code as the other stings to function for consistency.
- added new chem
- added new power
- added new ability
- added new disease
- added new human-species

## Why It's Good For The Game

Basically adds more slots for ghosts and unrevivable players/observers to use. This way they don't have to idle/wait as much as the game progresses and have something to do. This mechanic should also aid xenos in assaulting the FOB as these are expendable units that can help push. It's also a way to reward xenos for killing marines. Still way less of a bonus than infecting. (especially since you can reanimate bursted corpses) So it doesn't encourage intentionally permakilling marines instead of capturing. This also gives the hivelord another unique ability as a support caste to keep with the stinging theme for drone castes.

## Changelog
add: Necromorphs. Squishy, fast, expendable, ghost roles!
add: Hivelord sting. Reanimate the long dead, unrevivable marine corpses can now be turned into Necromorphs by the hivelord.
/:cl:

Mainly practicing the PR system. I've still to add some core content, such as new sprites.

### CODE REVIEW

**Comment by Ketrai** [comment]
Most errors are just because I have yet to include the new file into the DME.

**Comment by Ketrai** [comment]
So far everything works. can't sting corpses that haven't decayed etc. It just seems that the virus itself isn't working yet. Not really sure how to fix that.

Necromorphs themselves work as intended. Faster, squishier, don't infect, don't revive on death.

**Comment by Ketrai** [comment]
wut

**Comment by Ketrai** [comment]
No longer a complete WIP. This should be ready for testing and tweaking as I add more features later.

File: code/modules/mob/dead/observer/observer.dm [hunk 1]
```diff
@@ -598,6 +598,71 @@
 				return
 			SSticker.mode.spawn_larva(src, mother)
 
+/mob/dead/verb/join_as_necromorph()
+	set category = "Ghost"
+	set name = "Join as Necromorph"
+
+	var/mob/N = src
```
Comment (Rohesie) [inline code/modules/mob/dead/observer/observer.dm]: No need to type case this.

File: code/modules/mob/dead/observer/observer.dm [hunk 2]
```diff
@@ -598,6 +598,71 @@
...
+		return
+
+	if(jobban_isbanned(usr, "Alien"))
+		to_chat(usr, "<span class='warning'>You are banned from playing aliens and cannot spawn as a Necromorph.</span>")
+		return
+
+	var/list/necromorph_list = list()
+
+	for(var/mob/living/carbon/human/species/zombie/necromorph/A in GLOB.alive_mob_list)
+		if(istype(A) && !A.client)
```
Comment (Rohesie) [inline code/modules/mob/dead/observer/observer.dm]:
```suggestion
		if(!A.client)
```
No need to check for the type, as you are already doing that in the `for` lookup.

File: code/modules/mob/dead/observer/observer.dm [hunk 3]
```diff
@@ -598,6 +598,71 @@
...
+			necromorph_list += A.real_name
```
Comment (Rohesie) [inline code/modules/mob/dead/observer/observer.dm]: Why not adding a reference to the mob itself instead?

File: code/modules/mob/dead/observer/observer.dm [hunk 4]
```diff
@@ -598,6 +598,71 @@
...
+
+	if(!length(necromorph_list))
+		to_chat(usr, "<span class='warning'>There aren't any available Necromorphs.</span>")
+		return
+
+	var/choice = input("Pick a Necromorph:") as null|anything in necromorph_list
+	if(!choice)
+		return
+
+	for(var/mob/living/carbon/human/species/zombie/necromorph/P in GLOB.alive_mob_list)
```
Comment (Rohesie) [inline code/modules/mob/dead/observer/observer.dm]: Then we can skip this step.

File: code/modules/mob/dead/observer/observer.dm [hunk 5]
```diff
@@ -598,6 +598,71 @@
...
+
+	if(N.stat == DEAD)
+		to_chat(usr, "<span class='warning'>It's dead.</span>")
+		return
+
+	if(N.client)
+		to_chat(usr, "<span class='warning'>That player is still connected.</span>")
+		return
+
+	if(alert(usr, "Everything checks out. Are you sure you want to transfer yourself into this Necromorph?", "Confirmation", "Yes", "No") != "Yes")
```
Comment (Rohesie) [inline code/modules/mob/dead/observer/observer.dm]: The confirmation is unnecessary here, they can always cancel.

File: code/modules/mob/living/carbon/human/species.dm [hunk 6]
```diff
@@ -716,6 +716,67 @@
...
+	name_plural = "Necromorphs"
+	blood_color = "#1e0100"
+	slowdown = -1
+	brute_mod = 0.7
+	burn_mod = 0.9 //don't murderize me lasers
+	icobase = 'icons/mob/human_races/r_necro.dmi'
+	deform = 'icons/mob/human_races/r_necro.dmi'
+
+/datum/species/zombie/necromorph/handle_death(var/mob/living/carbon/human/H, gibbed)
+//	if(H && H.loc && H.stat == DEAD)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/human/species.dm]: You remove this comment.

File: code/modules/mob/living/carbon/human/species.dm [hunk 7]
```diff
@@ -716,6 +716,67 @@
...
+	H.gib()
+
+/datum/species/zombie/necromorph/handle_post_spawn(var/mob/living/carbon/human/H)
+	if(H.hud_used)
+		qdel(H.hud_used)
+		H.hud_used = null
+//		H.create_mob_hud()
+		if(H.hud_used)
+			H.hud_used.show_hud(H.hud_used.hud_version)
+	if(H.l_hand) H.dropItemToGround(H.l_hand, FALSE, TRUE)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/human/species.dm]: I know this is copypasted code, but some cleaning will be necessary.
Avoid same line control statements.

File: code/modules/mob/living/carbon/human/species.dm [hunk 8]
```diff
@@ -716,6 +716,67 @@
...
+	flags_item = NODROP|DELONDROP
+	force = 23
+	w_class = 4
+	sharp = 1
+	attack_verb = list("slashed", "bite", "tore", "scraped", "nibbled")
+	pry_capable = IS_PRY_CAPABLE_FORCE
+
+/obj/item/weapon/necromorph_claws/attack(mob/living/M, mob/living/carbon/human/user, def_zone)
+	if(user.species.count_human)
+		return 0
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/human/species.dm]:
```suggestion
		return FALSE
```

Comment (Rohesie) [reply code/modules/mob/living/carbon/human/species.dm]: This item shouldn't really be able to be picked by anyone, it should never happen naturally. This shouldn't be a thing.

File: code/modules/mob/living/carbon/human/species.dm [hunk 9]
```diff
@@ -716,6 +716,67 @@
...
+/datum/species/zombie/necromorph/handle_death(var/mob/living/carbon/human/H, gibbed)
+//	if(H && H.loc && H.stat == DEAD)
+	H.gib()
+
+/datum/species/zombie/necromorph/handle_post_spawn(var/mob/living/carbon/human/H)
+	if(H.hud_used)
+		qdel(H.hud_used)
+		H.hud_used = null
+//		H.create_mob_hud()
+		if(H.hud_used)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/human/species.dm]: This `if` can never be true as the code goes. Do you know what's going on here to keep it?

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 10]
```diff
@@ -2225,7 +2225,7 @@
 	to_chat(src, "<span class='xenodanger'>You feel your toxin glands refill, another young one ready for implantation. You can use Defile again.</span>")
 	update_action_button_icons()
 
-/mob/living/carbon/Xenomorph/proc/recurring_injection(mob/living/H, toxin1 = "xeno_toxin", toxin2 = null, larva = FALSE, count = 2)
+/mob/living/carbon/Xenomorph/proc/recurring_injection(mob/living/H, toxin1 = "xeno_toxin", toxin2 = null, larva = FALSE, virus1 = "necrovirus", count = 2)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]: Is there need to use this proc? You can make your own.

File: code/modules/mob/dead/observer/observer.dm (hunk 2 above)
Comment (Rohesie) [reply code/modules/mob/dead/observer/observer.dm]: Though ideally there's a global var of necromorphs instead. Same as for xenos or humans.

File: code/modules/reagents/chemistry_reagents/other.dm [hunk 11]
```diff
@@ -783,7 +783,7 @@
 		return
 	if(!(locate(/obj/effect/decal/cleanable/blackgoo) in T))
 		new /obj/effect/decal/cleanable/blackgoo(T)
-
+
```
Comment (Rohesie) [inline code/modules/reagents/chemistry_reagents/other.dm]: Avoid this kind of changes in unrelated files.

File: code/modules/mob/dead/observer/observer.dm [hunk 12]
```diff
@@ -598,6 +598,71 @@
...
+
+	if(SSticker.current_state < GAME_STATE_PLAYING)
+		to_chat(usr, "<span class='warning'>The game hasn't started yet!</span>")
+		return
+
+	if(jobban_isbanned(usr, "Alien"))
+		to_chat(usr, "<span class='warning'>You are banned from playing aliens and cannot spawn as a Necromorph.</span>")
+		return
+
+	var/list/necromorph_list = list()
```
Comment (spookydonut) [inline code/modules/mob/dead/observer/observer.dm]: make this a global list in `code/_globalvars/lists/mobs.dm`

File: code/modules/mob/living/carbon/xenomorph/Powers.dm (hunk 10 above)
Comment (Ketrai) [reply code/modules/mob/living/carbon/xenomorph/Powers.dm]: I suppose I can. Will make it more configurable. Will need a bit of time to figure it out though.

File: code/game/objects/effects/landmarks/corpsespawner.dm [hunk 13]
```diff
@@ -47,6 +47,7 @@
 	M.dna.mutantrace = mutantrace
 	M.real_name = name
 	M.death(1) //Kills the new mob
+	M.timeofdeath = 0
```
Comment (spookydonut) [inline code/game/objects/effects/landmarks/corpsespawner.dm]:
```suggestion
	M.status_flags |= NO_REANIMATE
```

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 14]
```diff
@@ -2428,3 +2440,59 @@
...
+	update_action_button_icons()
+
+/mob/proc/can_zombie()
+	return FALSE
+
+/mob/living/carbon/monkey/can_zombie()
+	return FALSE
+
+/mob/living/carbon/human/can_zombie()
+	if(stat == DEAD && world.time >= timeofdeath + CONFIG_GET(number/revive_grace_period))
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]:
```suggestion
	if(!(status_flags & NO_REANIMATE))
```

Comment (spookydonut) [reply code/modules/mob/living/carbon/xenomorph/Powers.dm]:
```suggestion
	if(stat == DEAD && !(status_flags & NO_REANIMATE) && world.time >= timeofdeath + CONFIG_GET(number/revive_grace_period))
```

File: code/modules/mob/living/carbon/human/species.dm (hunk 9 above)
Comment (Ketrai) [reply code/modules/mob/living/carbon/human/species.dm]: Nope, I don't really get what is going on there. It seems that it wants to delete huds humans are using i suppose? But the masks tend to get deleted, and huds are kind of tied to uh... glasses and such?

**Comment by Ketrai** [comment]
Think we could perhaps give this a small public test after which I can close this branch and look into building necros from scratch?  I just need to know if baseline necros add anything to the round or if they're just a bother and I shouldn't bother :P

File: code/__DEFINES/mobs.dm [hunk 15]
```diff
@@ -178,6 +178,7 @@ var/list/global_mutations = list() // list of hidden mutation things
 #define FAKEDEATH	8192	//Replaces stuff like changeling.changeling_fakedeath
 #define DISFIGURED	16384	//I'll probably move this elsewhere if I ever get wround to writing a bitflag mob-damage system
 #define XENO_HOST	32768	//Tracks whether we're gonna be a baby alien's mummy.
+#define NO_REANIMATE    (1 << 10)
```
Comment (DominikPanic) [inline code/__DEFINES/mobs.dm]: Change the other flags too please.

File: code/modules/mob/dead/observer/observer.dm [hunk 16]
```diff
@@ -598,6 +598,71 @@
...
+		return
+
+		if(N.client || N.stat == DEAD)
+			to_chat(usr, "<span class='warning'>Oops. That mob can no longer be controlled. Sorry.</span>")
+			return
+
+	var/mob/ghostmob = usr.client.mob
+	message_admins("[key_name(usr)] has joined as a [N].")
+	log_admin("[ADMIN_TPMONTY(usr)] has joined as a [N].")
+	N.ckey = usr.ckey
```
Comment (DominikPanic) [inline code/modules/mob/dead/observer/observer.dm]: Transfer the mind instead of changing the ckey and view like that.

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 17]
```diff
@@ -2241,6 +2241,18 @@
 		if(toxin2)
 			H.reagents.add_reagent(toxin2, DEFILER_STING_AMOUNT_RECURRING, null, 300, FALSE, FALSE, TRUE) //Caps the amount injected by the overdose limit
 			overdose_check(H, toxin2)
+		if(virus1)
+			var/necro_transforming = 0
+			if(!necro_transforming)
```
Comment (DominikPanic) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]: The above will always be false. What's the logic here?

**Comment by Ketrai** [comment]
Too complicated. Made a new PR instead #943 
Makes them into a caste, easier
